source/core/src/main/com/csse3200/game/inventory/Inventory.java

```diff
@@ -0,0 +1,303 @@
+     * @return the item at the specified index, or {@code null} if there is no item at that index
+     * @throws ArrayIndexOutOfBoundsException if the index is out of range, i.e,
+     *         {@code (index < 0 || index >= capacity)}
+     */
+    @Override
+    public AbstractItem getAt(int index) {
+        return inventory[index];
+    }
+
+    /**
+     * Deletes the first occurrence of an item with the specified code from the inventory.
+     * <p>Currently removes the item entirely. Needs to be updated to reduce quantity if
+     * applicable - ie if we only want to delete <b>n</b> of the item.</p>
+     *
+     * @param code the unique code of the item.
+     */
+    @Override
+    public void deleteItem(int code) {
+        this.getItemIndex(code).ifPresent(this::deleteItemAt);
+    }
+
+    /**
+     * Deletes the item at the specified index from the inventory.
+     * <p>It is currently up to the user to ensure they have not exceeded the bounds of the
+     * inventory.</p>
+     * <p>If the item is removed, updates the internal mappings and adjusts the next available
+     * index.</p>
+     *
+     * @param index the index of the item to delete.
+     */
+    @Override
+    public void deleteItemAt(int index) {
+        if (inventory[index] != null) {
+            this.removeAt(index);
+        }
+    }
+
+    /**
+     * Clears all items from the inventory.
+     * <p>Warning: All items will be deleted and cannot be recovered.</p>
+     */
+    @Override
+    public void clearInventory() {
+        freeSlots = capacity;
+        nextIndex = 0;
+        mapping.clear();
+        fill(inventory, null);
+    }
+
+    /**
+     * Uses the first occurrence of an item with the specified code in the inventory.
+     * <p>If the item is consumable and becomes empty after use, it is removed from the
+     * inventory.</p>
+     *
+     * @param code the unique code of the item.
+     * @param context the context in which the item is used (see {@link ItemUsageContext}).
+     */
+    @Override
+    public void useItem(int code, ItemUsageContext context) {
+        this.getItemIndex(code).ifPresent(index -> useAndPossiblyRemove(index, context));
+    }
+
+    /**
+     * Uses the item at the specified index in the inventory.
+     * <p>If the item is consumable and becomes empty after use, it is removed from the
+     * inventory.</p>
+     *
+     * @param index the index of the item to use.
+     * @param context the context in which the item is used (see {@link ItemUsageContext}).
+     */
+    @Override
+    public void useItemAt(int index, ItemUsageContext context) {
+        if (inventory[index] != null) {
+            this.useAndPossiblyRemove(index, context);
+        }
+    }
+
+    /**
+     * Adds an item to the inventory, either by stacking it with existing items or placing it in
+     * a new slot.
+     *
+     * @param item the item to add to the inventory.
+     */
+    @Override
+    public void add(AbstractItem item) {
```

Comment:
this should probably have a boolean return value to indicate whether pickup was successful; I expect items will behave differently if you try to pick them up and fail

Reply:
added